up-l1/mqtt_5.adoc

```diff
@@ -0,0 +1,135 @@
+
+For more information, please refer to https://mqtt.org/
+
+This document will discuss the uTransport implementation on MQTT 5.
+
+== Specifications
+
+=== UAttribute Mapping
+
+MQTT 5 supports custom header fields, and we leverage this to map the UAttribute values into the MQTT header.
+
+[cols="1,1"]
+|===
+| MQTT Header Property Code | Value (key:value)
+| UserProperty
+| ("1", "{UAttribute.id}") # UUID id
+| UserProperty
+| ("2", "{UAttribute.type}") # UMessageType
+| UserProperty
+| ("3","{UAttribute.source}") # Short Form Serialized Source UURI
+| UserProperty
+| ("4","{UAttribute.sink}") # Short Form Serialized Sink UURI
+| UserProperty
+| ("5","{UAttribute.priority}") # UPriority
+| UserProperty
+| ("6", "{UAttribute.ttl}") # Optional ttl
+| UserProperty
+| ("7", "{UAttribute.permissionLevel}") # Optional permission level
+| UserProperty
+| ("8", "{UAttribute.commStatus}") # Optional UCode commstatus
+| UserProperty
+| ("9", "{UAttribute.reqId}") # UUID reqid
+| UserProperty
+| ("10", "{UAttribute.token}") # Optional token
+| UserProperty
+| ("11", "{UAttribute.traceParent}") # Optional traceparent
+|===
+
+All uAttributes are mapped to a MQTT header UserProperty, where the key is the UAttribute protobuf field number. The value is a string representation of the UAttribute field.
+
+=== URI Mapping to MQTT topics
+
+The MQTT topic a message is published on utilizes elements from both the source and sink UUri fields. Which elements from the UUri are used in the topic depend on whether the sink UUri is present. In addition, there is a client identifier that defines where the uTransport client is located.
+
+Has sink UUri:
+
+`{client_identifier}/{source.authority_name}/{sink.authority_name}/{sink.ue_id}/{sink.ue_version_major}/{sink.resource_id}`
+
+No sink UUri (Publish message type only):
+
+`{client_identifier}/{source.authority_name}//{source.ue_id}/{source.ue_version_major}/{source.resource_id}`
+
+The entity and resource elements always represent the destination of the message. In the case of a Publish message type, the destination happens to be based on the source UUri. The source authority information is also always present to allow for authorization controls based on the authority information, since an authority maps to the uE sending the message.
+
+The client_identifier is used to determine where messages are coming from. The valid values for the client_identifier are:
+
+- `c` : Indicates that the message originated from the cloud
+- `d` : Indicates that the message originated from a device and is remote
+- `l` : Indicates that the message originated from a device and is local to that device
+
+The client_identifier that is used for the topic is based on the kind of MQTT uTransport client that is sending the message. An MQTT uTransport client can be one of the following:
```

Comment:
Hey @devkelley -- following up on this from our talk this morning.

Is the client_identifier something that would be set in, say, the constructor of `UPClientMQTT5`? I'm trying to understand where this information comes from so that when an MQTT5 topic name is crafted within `UPClientMQTT5`, the correct client_identifier is prepended.

Reply:
Correct, this would be initialized by the constructor of the UPClientMQTT5. If a client was initialized on a device then `d` would be prepended, and if a client was initialized in the cloud then `c` would be prepended.